File: 0395-longest-substring-with-at-least-k-repeating-characters/0395-longest-substring-with-at-least-k-repeating-characters.java
class Solution {
    public int longestSubstring(String s, int k) {
        if(s==null || k>s.length())return 0;
        
        int n=s.length();
        
        char []chars=new char[26];
        
        for(char c:s.toCharArray())chars[c-'a']++;
        
        int idx=0;
        
        while(idx<n && chars[s.charAt(idx)-'a']>=k )idx++;
        
        if(idx==n)return n;
        
        int leftSubarray=longestSubstring(s.substring(0,idx),k);
        
        while(idx<n && chars[s.charAt(idx)-'a']<k )idx++;
        
        int rightSubarray=longestSubstring(s.substring(idx),k);
        
        return Math.max(leftSubarray,rightSubarray);
        
    }
}

File: 0395-longest-substring-with-at-least-k-repeating-characters/0395-longest-substring-with-at-least-k-repeating-characters.cpp
class Solution {
public:
    int longestSubstring(string s, int k) {
        
        if(s.size()==0 || k>s.size())return 0;
        
        unordered_map<char,int>map;
        for(char c:s)map[c]++;
        int i=0;
        
        while(i<s.size() && map[s[i]]>=k)i++;
 
        
        if(i==s.size())return i;
        
        int leftSubarray=longestSubstring(s.substr(0,i),k);
               
        while(i<s.size() && map[s[i]]<k)i++;
        
        int rightSubarray=longestSubstring(s.substr(i),k);
        
        return max(leftSubarray,rightSubarray);

        
    }
};